extensions/amp-next-page/0.2/service.js

```diff
@@ -227,17 +220,24 @@ export class NextPageService {
   }
 
   /**
-   * @return {!PositionObserver}
-   * @private
+   * Makes sure that all pages preceding the current page are
+   * marked hidden if they are out of the viewport
+   * @param {number} index index of the page to start at
    */
-  getPositionObserver_() {
-    // For testing
-    if (this.injectedPositionObserver_) {
-      return this.injectedPositionObserver_;
+  hidePreviousPages(index) {
+    let i = index + this.scrollDirection_;
```

Comment:
Would consider changing this logic to something like:

```javascript
const pagesToHide = this.scrollDirection > 0 ? this.pages_.slice(index + 1) : this.pages_.slice(0, index);
pagesToHide.filter(page => {
  const shouldBeHidden =
        page.relativePos === ViewportRelativePos.LEAVING_VIEWPORT ||
        page.relativePos === ViewportRelativePos.OUTSIDE_VIEWPORT ||
        page === this.initialPage_;
   return shouldBeHidden && page.isVisible();
})
.forEach(page => page.setVisibility(VisibilityState.HIDDEN));
```

The benefit here would be the logic for getting the candidate pages to hide, figuring out which ones need to hide and actually hiding are no longer intermixed.